addons/l10n_es_edi_sii/models/l10n_es_edi_certificate.py

```diff
@@ -0,0 +1,87 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from base64 import b64decode
+from pytz import timezone
+from datetime import datetime
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.serialization import Encoding, NoEncryption, PrivateFormat, pkcs12
+
+
+from odoo import _, api, fields, models, tools
+from odoo.exceptions import ValidationError
+
+
+class Certificate(models.Model):
+    _name = 'l10n_es_edi.certificate'
+    _description = 'Personal Digital Certificate'
+    _order = 'date_start desc, id desc'
+    _rec_name = 'date_start'
+
+    content = fields.Binary(string="File", required=True, help="PFX Certificate")
+    password = fields.Char(help="Passphrase for the PFX certificate", groups="base.group_system")
+    date_start = fields.Datetime(readonly=True, help="The date on which the certificate starts to be valid")
+    date_end = fields.Datetime(readonly=True, help="The date on which the certificate expires")
+    company_id = fields.Many2one(comodel_name='res.company', required=True, default=lambda self: self.env.company)
+
+    # -------------------------------------------------------------------------
+    # HELPERS
+    # -------------------------------------------------------------------------
+
+    @api.model
+    def _get_es_current_datetime(self):
+        """Get the current datetime with the Peruvian timezone. """
+        return datetime.now(timezone('Europe/Madrid'))
+
+    @tools.ormcache('self.content', 'self.password')
+    def _decode_certificate(self):
+        """Return the content (DER encoded) and the certificate decrypted based in the point 3.1 from the RS 097-2012
+        http://www.vauxoo.com/r/manualdeautorizacion#page=21
+        """
+        self.ensure_one()
+
+        if not self.password:
+            return None, None, None
+
+        private_key, certificate, dummy = pkcs12.load_key_and_certificates(
+            b64decode(self.content),
+            self.password.encode(),
+            backend=default_backend(),
+        )
+
+        pem_certificate = certificate.public_bytes(Encoding.PEM)
+        pem_private_key = private_key.private_bytes(
+            Encoding.PEM,
+            format=PrivateFormat.TraditionalOpenSSL,
+            encryption_algorithm=NoEncryption(),
+        )
+        return pem_certificate, pem_private_key, certificate
+
+    # -------------------------------------------------------------------------
+    # LOW-LEVEL METHODS
+    # -------------------------------------------------------------------------
+
+    @api.model
+    def create(self, vals):
+        record = super().create(vals)
+
+        spain_tz = timezone('Europe/Madrid')
+        spain_dt = self._get_es_current_datetime()
+        try:
+            dummy, dummy, certificate = record._decode_certificate()
```

Comment:
Prefix good name with `_`, don't use garbage names.

Reply:
not `_`. it is used below to translate the error message

Reply:
I can't because `_` is already the translation method...

Reply:
Note: it could be nice to rename the translation method by `_t` or something like that to unlock anonymous variables

Reply:
> not `_`. it is used below to translate the error message

> I can't because `_` is already the translation method...

**prefix**, not replace. In the previous pr the return values were called `pem_certificate, pem_private_key, certificate` but ci/style complains because the first two are unused.

Renaming them to `_pem_certificate` and `_pem_private_key` would have made pyling happy, keeping the code readable instead of getting nonsense.

> Note: it could be nice to rename the translation method to `_t` or something like that to unlock anonymous variables

Just prefix them with `_`. Or use `__` (double underscore). Or don't, that looks ugly.

Reply:
I didn't know about prefixing the variable by `_` was working to satisfy pylint... I agree this is far more understandable than the `dummy` variable